### ISSUE TITLE
thread: outline of per-VCI thread model

### ISSUE THREAD

**Issue by hzhou** [pmodels/mpich#4219]
This issue outlines the work to be done for the per-VCI thread model. 

It is a work-in-progress.
* [x] Finalize a clean set of TODO/check list.

## Background
* Previous work separates the communication path from the global lock scheme and uses a global vci lock
* PR #3921 Further splits the global vci lock by "VCI" that is assigned to each communicator. 

## Two main components
There are two main components for thread safety: communication path and non-communication path.

Non-communication path is non-performance critical, and it can benefit from coarse thread granularity -- simpler code as well as better serial performance due to fewer locks.

The communication path is performance-critical and incurring unnecessary serialization is detrimental. This path needs fine-grained critical sections -- typically tight lock-unlock blocks, or even atomics, around netmod API calls.

ref: #4216 

## Communication operation paths
There are 3 main communication paths: pt2pt, collective, and RMA. pt2pt and RMA paths are similar, both can/should be further distinguished into direct netmod path, active message path, shmmod path

### Pt2pt path
In general, this path only takes on fine-grained locks. This path should not take on any coarse locks to prevent inadvertent serialization or (worse) deadlocking.
 
Fine-grained locks include:
* per-VCI/VNI locks for netmod
* per-object locks for fbox, iqueue etc. for the shmmod
* per-thread cache or atomics or per-object locks for request allocation/deallocation.

We should try to avoid yield unless it is yielding within the context. Only yield a lock that is explicitly taken within the same function. When the lockes are designed sufficiently fine-grained, yielding should be necessary. However, for performance-critical progress functions (e.g. `Waitall`), coarser grain may be desirable. Even in this case, we should strive to keep the locks within context -- duplicating safe/unsafe versions of functions if necessary.

### Collective path
Collectives cannot overlap by sematics. It will always decompose into pt2pt operations which does not interfere with coarse locks. Therefore, I think Collectives can benefit from a single coarse lock, either the "global" lock or a collective lock. The former will serialize collectives with other MPI operations such as datatype and communicator operations. Thos operations are very heavy. Therefore, it is desirable for collectives to take on its own per-object lock.
* [ ] create PR to let collectives take `MPID_THREAD_VCI_COLLECTIVE_MUTEX`

### Request creation etc.
* [ ] Thread-safety of request objects via atomics and per-thread-cache
* [ ] Alternative per-vci thread pool will require individual lock

### RMA path
RMA path is similar to pt2pt path. Therefore, the same bullets apply here.

* [ ] Add granular lock on active message protocol data structure
* [ ] VNI lock on netmod function calls

### Direct netmod path
Tight lock-unlock vni mutex, essentially making netmod operation atomic.

Essential information are all encapsulated in a request object, which itself needs protection.

### Active message path
Active message path are generally of two layers, the upper protocol layer and the lower communication layer. The communication layer for active messages are independent in nature, therefore can be made in parallel, for example, distribute across vci channels. 

The protocol layer are being accessed asynchronously. Each protocol may require its own per-protocol locking to prevent corruptions.

* [ ] Add granular lock on generic queues (e.g. `posted_list`, `unexpected_list`, etc.)

### SHMMOD path
Identify the data structure and create per-object locks for each (global) data structure.

Currently shmmod follows the active message path.

## Communication progress paths
The progress paths are distinct in that:
* Implicit. While user can provoke progress explicitly with, e.g. `MPI_Wait`, it is implicit in which operation it progresses.
* Cross-component. Potentially it may touch multiple vci and both send and recv data structures.
* Non-deterministic. It is not certain how much progress will be made. It will invoke dispatch functions which vary both in term of timing and in term of components.

While the operation paths can benefit from pushing critical section down to device layer, the progress path can benefit from coarser per-component critical sections.

## Non-critical communication paths
These paths can either use global lock or granular locks. Either way, we need ensure to unlock before calling communication routines. 
* [ ] Communication/context_id creations
* [ ] Dynamic process connections

**Comment by wesbland** [comment]
cc: @sagarth

**Comment by hzhou** [comment]
## Current status
Current per-vci thread model roughly separate communication path with a coarse lock -- `MPIDI_global.vci_lock` and the rest of the path with `GLOBAL_ALLFUNC_MUTEX` (to be renamed to `VCI_GLOBAL_MUTEX`. The communication path can be further resolved into send path, recv path and progress path.

There are two dominant non-fast path components that intervenes with the communication path: context_id generation during communicator creation and connection establishing during dynamic process creation. Both of these two components need use their own per-object locks.

**Comment by hzhou** [comment]
superseded by #4649
